api/trace/b3_propagator.go

```diff
@@ -34,185 +34,295 @@ const (
+	switch strings.ToLower(sampled) {
+	case "0", "false":
+		// Zero value for TraceFlags sample bit is unset.
+	case "1", "true":
+		sc.TraceFlags = FlagsSampled
+	case "":
+		sc.TraceFlags = FlagsDeferred
+	default:
+		return empty, errInvalidSampledHeader
 	}
-	sampled, ok := b3.extractSampledState(supplier.Get(B3SampledHeader))
-	if !ok {
-		return EmptySpanContext()
+
+	// The only accepted value for Flags is "1". This will set Debug to
+	// true. All other values and omission of header will be ignored.
+	if flags == "1" {
+		sc.TraceFlags |= FlagsDebug
 	}
 
-	debug, ok := b3.extracDebugFlag(supplier.Get(B3DebugFlagHeader))
-	if !ok {
-		return EmptySpanContext()
+	if traceID != "" {
+		requiredCount++
+		id := traceID
+		if len(traceID) == 16 {
+			// Pad 64-bit trace IDs.
+			id = b3TraceIDPadding + traceID
+		}
+		if sc.TraceID, err = IDFromHex(id); err != nil {
+			return empty, errInvalidTraceIDHeader
+		}
 	}
-	if debug == FlagsSampled {
-		sampled = FlagsSampled
+
+	if spanID != "" {
+		requiredCount++
+		if sc.SpanID, err = SpanIDFromHex(spanID); err != nil {
+			return empty, errInvalidSpanIDHeader
+		}
 	}
 
-	sc := SpanContext{
-		TraceID:    tid,
-		SpanID:     sid,
-		TraceFlags: sampled,
+	if requiredCount != 0 && requiredCount != 2 {
+		return empty, errInvalidScope
 	}
 
-	if !sc.IsValid() {
-		return EmptySpanContext()
+	if parentSpanID != "" {
+		if requiredCount == 0 {
+			return empty, errInvalidScopeParent
+		}
+		// Validate parent span ID but we do not use it so do not save it.
+		if _, err = SpanIDFromHex(parentSpanID); err != nil {
+			return empty, errInvalidParentSpanIDHeader
+		}
 	}
 
-	return sc
+	return sc, nil
 }
 
-func (b3 B3) extractSingleHeader(supplier propagation.HTTPSupplier) SpanContext {
-	h := supplier.Get(B3SingleHeader)
-	if h == "" || h == "0" {
-		return EmptySpanContext()
-	}
-	sc := SpanContext{}
-	parts := strings.Split(h, "-")
-	l := len(parts)
-	if l > 4 {
-		return EmptySpanContext()
-	}
-
-	if l < 2 {
-		return EmptySpanContext()
+// extractSingle reconstructs a SpanContext from contextHeader based on a B3
+// Single header. It is based on the implementation found here:
+// https://github.com/openzipkin/zipkin-go/blob/v0.2.2/propagation/b3/spancontext.go
+// and adapted to support a SpanContext.
+func extractSingle(contextHeader string) (SpanContext, error) {
+	if contextHeader == "" {
+		return empty, errEmptyContext
 	}
 
-	var err error
-	sc.TraceID, err = IDFromHex(fixB3TID(parts[0]))
-	if err != nil {
-		return EmptySpanContext()
-	}
+	var (
+		sc       = SpanContext{}
+		sampling string
+	)
 
-	sc.SpanID, err = SpanIDFromHex(parts[1])
-	if err != nil {
-		return EmptySpanContext()
-	}
+	headerLen := len(contextHeader)
 
-	if l > 2 {
-		var ok bool
-		sc.TraceFlags, ok = b3.extractSampledState(parts[2])
-		if !ok {
-			return EmptySpanContext()
+	if headerLen == 1 {
+		sampling = contextHeader
+	} else if headerLen == 16 || headerLen == 32 {
```

Comment:
I guess we don't really use `nibble` much these days, but if we wanted to avoid magic constants, we could do something like

```go
const (
    NibblesIn64BitHeader = 64/4
    NibblesIn128BitHeader = 128/4
)
```

Reply:
Updated to name all field widths.